File: my-folder/0046-permutations/solution.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> list = new ArrayList<>();
        helper(res, list, nums);
        return res;
    }
    
    private void helper(List<List<Integer>> res, List<Integer> list, int[] nums) {
        //base case
        if (list.size() == nums.length) {
            res.add(new ArrayList<Integer>(list));
            return;
        }
        
        for (int i = 0; i < nums.length; i++) {
            if (list.contains(nums[i])) continue;
            list.add(nums[i]);
            helper(res, list, nums);
            list.remove(list.size() - 1);
        }
    }
}

File: my-folder/0046-permutations/solution.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        used = [False]*len(nums)
        now_res = []
        def traverse(now_res):
            if len(now_res)==len(nums):
                res.append(now_res)
            else:
                for i in range(len(nums)):
                    if used[i]!=True:
                        used[i]=True
                        traverse(now_res+[nums[i]])
                        used[i]=False
                
        traverse([])
        return res

        
